graphql_compiler/query_formatting/common.py

```diff
@@ -34,91 +35,84 @@
-    expected_python_types = allowed_types_for_graphql_type[expected_type.name]
-    if isinstance(value, bool) and not is_same_type(GraphQLBoolean, expected_type):
-        correct_type = False  # We explicitly disallow passing boolean values for non-boolean types
-    if not isinstance(value, expected_python_types):
-        correct_type = False
-    if not correct_type:
-        raise ValueError(
-            "Unexpected type {}. Expected one of {}.".format(type(value), expected_python_types)
+        deserialization_function = _CUSTOM_SCALAR_DESERIALIZATION_FUNCTIONS.get(
+            graphql_type.name, graphql_type.parse_value
         )
+        try:
+            return deserialization_function(value)
+        except (ValueError, TypeError) as e:
+            raise GraphQLInvalidArgumentError("Error parsing argument {}: {}".format(name, e))
+
+    return _json_scalar_deserialization_function
+
+
+_ALLOWED_JSON_SCALAR_TYPES = MappingProxyType(
+    {
+        GraphQLDate: (str,),
+        GraphQLDateTime: (str,),
+        GraphQLFloat: (str, float, int),
+        GraphQLDecimal: (str, float, int),
+        GraphQLInt: (int, str),
+        GraphQLString: (str,),
+        GraphQLBoolean: (bool,),
+        GraphQLID: (int, str,),
+    }
+)
+# In general, we should compare two GraphQLScalarType objects by using their names since different
+# GraphQLScalarType objects can refer to the same type. However, in this case we can and should use
+# direct equality to ensure that the parsing and serialization functions are also the same.
+assert_set_equality(set(_ALLOWED_JSON_SCALAR_TYPES.keys()), SUPPORTED_SCALAR_TYPES)
+
+_SCALAR_DESERIALIZATION_FUNCTIONS = MappingProxyType(
+    {
+        graphql_type.name: _get_json_scalar_deserialization_function(graphql_type, allowed_types)
+        for graphql_type, allowed_types in _ALLOWED_JSON_SCALAR_TYPES.items()
+    }
+)
 
-    name_to_custom_type = {graphql_type.name: graphql_type for graphql_type in CUSTOM_SCALAR_TYPES}
-    # Parse the value. In most cases we can use the default GraphQL parser, but there are some
-    # special cases where we are more permissive than the default.
-    # By default, strings cannot be parsed to float.
-    if is_same_type(expected_type, GraphQLFloat):
-        return float(value)
-    # Use the default GraphQL parser to parse the value
-    elif expected_type.name in name_to_custom_type:
-        # Since we cannot serialize the parse_value function of custom scalar types when
-        # serializing a schema, it is possible that the parse_value is incorrectly set. We must,
-        # therefore, use the parse_value function of the original scalar type definition.
-        return name_to_custom_type[expected_type.name].parse_value(value)
-    else:
-        return expected_type.parse_value(value)
+
+def _deserialize_json_scalar_argument(name, expected_type: GraphQLScalarType, value: Any) -> Any:
+    """Deserialize the json serialized scalar argument."""
```

Comment:
Can you go through all the docstrings in this module (including the ones you merged in prior PRs) and reword them to be more descriptive and helpful for an unknowledgeable reader? Right now, many of them (e.g. this one) look like this:
```python
def do_thing():
    """Do the thing"""
```
This docstring provides no additional useful information that couldn't already be inferred from the function name. After reading this, I probably don't know what the function was actually intended for, what problem it solves, and when I should or should not be calling it.

Please avoid docstrings like that. It's better to not have docstrings than to have completely unhelpful ones.